### PR TITLE
Add Time condition to Conditional Logger

### PR DESCRIPTION
**Pull request by apaliy**
Add Timeout logger for CB open method.

### CODE REVIEW

File: src/main/java/org/prebid/server/log/ConditionalLogger.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                onLog.accept(logger);
+            }
+            return result;
+        });
+    }
+
+    private AtomicLong recalculateDate(long amount, TimeUnit unit) {
+        final long amountInMillis = unit.toMillis(amount);
+        final Date result = new Date(currentTimeMillis() + amountInMillis);
+        return new AtomicLong(result.getTime());
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/log/ConditionalLogger.java]: Try to use `Instant.ofEpochMilli()` probably

Comment (DGarbar) [inline src/main/java/org/prebid/server/log/ConditionalLogger.java]: Not sure about `AtomicLong` usage here. 
You don't change variable inside it, you recreating and replacing value in `ConcurrentHashMap` each time

File: src/test/java/org/prebid/server/log/ConditionalLoggerTest.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+@RunWith(MockitoJUnitRunner.class)
+public class ConditionalLoggerTest {
+
+    @Mock
+    private Logger logger;
+
+    private ConditionalLogger conditionalLogger;
+
+    @Before
+    public void setUp() throws Exception {
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/log/ConditionalLoggerTest.java]: Is throws Exception necessary ?

File: src/test/java/org/prebid/server/log/ConditionalLoggerTest.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    @Test
+    public void log() {
+        //when
+        for (int i = 0; i < 100; i++) {
+            conditionalLogger.info("Hello", 20);
+        }
+        //then
+        verify(logger, times(5)).info(any());
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/log/ConditionalLoggerTest.java]: Change any() to "Hello"

File: src/main/java/org/prebid/server/metric/MetricName.java [hunk 4]
```diff
@@ -75,7 +75,10 @@
 
     // cache
     prebid_cache_request_success_time,
-    prebid_cache_request_error_time;
+    prebid_cache_request_error_time,
+
+    //account.*.requests.
+    rejected;
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/metric/MetricName.java]: Update `metrics.md`

**Comment by apaliy** [comment]
Sorry for prev commit message(

File: src/main/java/org/prebid/server/log/ConditionalLogger.java [hunk 5]
```diff
@@ -40,4 +62,25 @@ private void log(String key, Integer maxValue, Consumer<Logger> consumer) {
...
+    }
+
+    private long recalculateDate(long amount, TimeUnit unit) {
+        final long amountInMillis = unit.toMillis(amount);
+        Instant resultInstant = Instant.now().plusMillis(amountInMillis);
+        return resultInstant.toEpochMilli();
+    }
+
+    private long currentTimeMillis() {
+        return new Date().getTime();
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/log/ConditionalLogger.java]: Try to migrate from old `Data` to newer `.time.` library please

File: src/test/java/org/prebid/server/log/ConditionalLoggerTest.java [hunk 6]
```diff
@@ -23,17 +22,17 @@
...
-    public void setUp() throws Exception {
+    public void setUp() {
         conditionalLogger = new ConditionalLogger(logger);
     }
 
     @Test
-    public void log() {
-        // when
+    public void counterLog() {
+        //when
```
Comment (rpanchyk) [inline src/test/java/org/prebid/server/log/ConditionalLoggerTest.java]: Minor. Why to skip space? :)

File: src/main/java/org/prebid/server/log/ConditionalLogger.java [hunk 7]
```diff
@@ -1,37 +1,67 @@
...
+    private ConcurrentMap<String, AtomicInteger> messageToCount;
+    private ConcurrentMap<String, Long> messageToWait;
+
     private final Logger logger;
 
     public ConditionalLogger(Logger logger) {
         this.logger = logger;
-        this.messageToCount = new ConcurrentHashMap<>();
+        this.messageToWait = Caffeine.newBuilder()
+                .maximumSize(10_000)
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/log/ConditionalLogger.java]: Pls extract `10_000` to class constant.

File: src/main/java/org/prebid/server/log/ConditionalLogger.java [hunk 8]
```diff
@@ -1,37 +1,67 @@
...
+                .expireAfterWrite(1, TimeUnit.HOURS)
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/log/ConditionalLogger.java]: Pls extract `1` to class constant.

File: src/main/java/org/prebid/server/handler/openrtb2/AuctionHandler.java [hunk 9]
```diff
@@ -160,7 +160,7 @@ private void handleResult(AsyncResult<Tuple2<BidResponse, AuctionContext>> respo
 
                 status = HttpResponseStatus.UNAUTHORIZED.code();
                 body = message;
-                String userId = ((UnauthorizedAccountException) exception).getAccountId();
+                final String userId = ((UnauthorizedAccountException) exception).getAccountId();
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/handler/openrtb2/AuctionHandler.java]: Pls rename to `accountId`.

File: src/main/java/org/prebid/server/geolocation/CircuitBreakerSecuredGeoLocationService.java [hunk 10]
```diff
@@ -40,7 +43,7 @@ public CircuitBreakerSecuredGeoLocationService(Vertx vertx, GeoLocationService g
     }
 
     private void circuitOpened() {
-        logger.warn("GeoLocation service is unavailable, circuit opened.");
+        conditionalLogger.warn("GeoLocation service is unavailable, circuit opened.", 5, TimeUnit.SECONDS);
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/geolocation/CircuitBreakerSecuredGeoLocationService.java]: Pls extract `5` to class constant, like `LOG_PERIOD_SECONDS`.

File: src/main/java/org/prebid/server/log/ConditionalLogger.java [hunk 11]
```diff
@@ -40,4 +70,25 @@ private void log(String key, Integer maxValue, Consumer<Logger> consumer) {
             consumer.accept(logger);
         }
     }
+
+    public void log(String key, long amount, TimeUnit unit, Consumer<Logger> consumer) {
+        messageToWait.compute(key, (currentKey, lastTimeMillis) -> {
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/log/ConditionalLogger.java]: Do we really need to `compute` value at each execution?

**Pull request merged**